### ISSUE TITLE
Proposal for extensible input/field types from the JSON block definition

### ISSUE THREAD

**Issue by jollytoad** [google/blockly#1584]
### Problem statement

Adding or modifying an input or field type in the block JSON definition requires customisation of the `Blockly.Block.prototype.interpolate_` function - adding or changing a `case` in the `switch` statement ... https://github.com/google/blockly/blob/master/core/block.js#L1143

### Proposal

I'd like to suggest replacing the switch with a object of input/field handlers, into which we can add new types or override existing types.

```javascript
Blockly.Block.jsonInputTypes = {
  input_value: function(element) {
    return this.appendValueInput(element['name']);
  },
  ...
};

Blockly.Block.jsonFieldTypes = {
  field_label: function(element) {
    return Blockly.Block.newFieldLabelFromJson_(element);
  },
  ...
};
```

and replace the switch with:

```javascript
var inputType = Blockly.Block.jsonInputTypes[element['type']]
var fieldType = Blockly.Block.jsonFieldTypes[element['type']]

if (inputType) {
  input = inputType.call(this, element)
} else if (fieldType) {
  field = fieldType.call(this, element)
}

// The input/field fn may return null so it falls back to the 'alt' definition

if (!input && !field && element['alt']) {
  element = element['alt'];
  altRepeat = true;
}
```

if its something that you think could be accepted into the codebase I'd be more than happy to prepare a PR and sign the CLA.

**Comment by rachel-fenichel** [comment]
Definitely a good idea to do this for fields, and we'd be happy to have a PR for that.  

My inclination is to leave inputs alone, with the few hard-coded types, because adding a new input type is not something we see people doing as frequently.  Do you have a use case for that in particular?

@NeilFraser do you have an opinion on whether this should be done by writing directly to a `Blockly.Block.fieldJsonTypes` object, vs registering the function some other way?

**Comment by RoboErikG** [comment]
I have a preference for API calls so that it's less painful if we need to change the implementation in the future.

**Comment by jollytoad** [comment]
Yeah I was thinking an API call to register/override rather than just assigning to the map, didn't want to overwhelm the proposal with too many details.

I included inputs as they are already handled in the same fn as fields, and it seemed appropriate to handle them in the same way and ditch the switch entirely rather than having some things handled in switch and others delegating.

I do actually have a use case for a custom input - i've designed some blocks where the blocks within an input_statement become elements in a list (actually lambda fns processed by the parent) - I've had to do some seriously dirty hacking in the code generator to support this - whereas a custom input would make it a lot easier. Although there are other areas in the codebase that need hacking to support custom inputs, but i think allowing custom inputs here would be a good first step towards that.

I'll knock together an initial PR and the details can be debated then.

**Comment by jollytoad** [comment]
I've just come across this... https://github.com/google/blockly/issues/1486

Wondering if it may be worth combining the effort here, so the mapping becomes one from the type string to class (eg. 'field_label' -> Blockly.FieldLabel), and we just call fromJson on that?

**Comment by jollytoad** [comment]
Just taken a quick look at the input code, and to be clear, I think i'd leave inputs alone for now if doing the work in conjunction with #1486

**Comment by jollytoad** [comment]
I've implemented this on top of #1486, i'm not sure how to present a PR based on an another PR (or even if its possible) in Github so i've created it in my own fork for now until #1592 is merged (if accepted) ... you can review the potential PR here... https://github.com/jollytoad/blockly/pull/1/files

**Comment by jollytoad** [comment]
@rachel-fenichel thanks for the quick turn around on #1592, now that's merged I've submitted a proper PR for this ^
